cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+{
+    assert(package_module && package_module->name);
+
+    PackageModuleWrapper *wrapper = xmalloc(sizeof(PackageModuleWrapper));
+    
+    wrapper->path = GetPackageModuleRealPath(package_module->name);
+    wrapper->name = SafeStringDuplicate(package_module->name);
+    wrapper->package_module = package_module;
+    
+    /* Check if file exists */
+    struct stat sb;
+    if (!wrapper->path || (stat(wrapper->path, &sb) != 0))
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "can not find package wrapper in provided location: %s",
+            wrapper->path);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    
+    /* Negotiate API version */
+    wrapper->supported_api_version = NegotiateSupportedAPIVersion(wrapper);
+    if (wrapper->supported_api_version != 1)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "unsupported package module wrapper API version: %d",
+            wrapper->supported_api_version);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    return wrapper;
+}
+
+
+void UpdatePackagesCache(EvalContext *ctx, bool force_update)
+{
+    Log(LOG_LEVEL_DEBUG, "Updating package cache.");
+    
+    PackagePromiseGlobalLock package_lock = 
+            AcquireGlobalPackagePromiseLock(ctx);
+                        
+    if (package_lock.g_lock.lock == NULL)
+    {
+        Log(LOG_LEVEL_INFO, "Can not acquire global lock for package promise. "
+            "Skipping updating cache.");
+        return;
+    }
+                        
+    Rlist *default_inventory = GetDefaultInventoryFromContext(ctx);
+    
+    for (const Rlist *rp = default_inventory; rp != NULL; rp = rp->next)
+    {
+        const char *pm_name =  RlistScalarValue(rp);
+        
+        /* We don't want inventory to be reported. */
+        if (StringSafeEqual(pm_name, "cf_null"))
+        {
+            break;
+        }
+        
+        PackageModuleBody *module = GetPackageModuleFromContext(ctx, pm_name);
+        if (!module)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Can not find body for package module: %s", pm_name);
+            continue;
+        }
+
+        PackageModuleWrapper *module_wrapper = NewPackageModuleWrapper(module);
+
+        if (!module_wrapper)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Can not set up wrapper for module: %s", pm_name);
+            continue;
+        }
+
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       UPDATE_TYPE_INSTALLED, force_update);
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       force_update ? UPDATE_TYPE_UPDATES : 
+                                           UPDATE_TYPE_LOCAL_UPDATES,
+                                       force_update);
+
+        DeletePackageModuleWrapper(module_wrapper);
+    }
+    YieldGlobalPackagePromiseLock(package_lock);
```

Comment:
No. default_inventory is taken from ctx which is persistent for the whole agent run and is destroyed at the end of its run.